File: frontend/src/lib/deployGate.js
/** Deploy gate — forward-test before capital (mirrors backend deploy_gate.py). */

import { backtestFingerprint } from './backtestDisplay';
import { formatDirectionModeLabel, normalizeDirectionMode } from './botConfigDisplay';

export { normalizeDirectionMode, formatDirectionModeLabel };

export const DEPLOY_GATE_DEFAULTS = {
  minTrades: 1,
  minPnl: 0,
  minStabilityScore: 0.5,
  maxDrawdownWarnPct: 25,
};

function check({ id, level, ok, message, detail }) {
  return { id, level, ok, message, detail };
}

/** sim_mode from backtest wire payload (top-level or meta.config). */
export function extractBacktestSimMode(results) {
  if (!results) return 'live_aligned';
  const cfg = results.meta?.config ?? {};
  return String(results.sim_mode ?? cfg.sim_mode ?? 'live_aligned').toLowerCase();
}

/** direction_mode used for the backtest run (meta.config, sweep best, or explicit override). */
export function extractBacktestDirectionMode(results, backtestConfig) {
  if (backtestConfig?.direction_mode != null) {
    return normalizeDirectionMode(backtestConfig.direction_mode);
  }
  const cfg = results?.meta?.config ?? {};
  if (cfg.direction_mode != null) {
    return normalizeDirectionMode(cfg.direction_mode);
  }
  const best = results?.sweep?.best?.config;
  if (best?.direction_mode != null) {
    return normalizeDirectionMode(best.direction_mode);
  }
  return 'LONG_ONLY';
}

function symbolSlice(results, symbol) {
  if (!results?.portfolio || !symbol) return { scoped: results, symbolError: null };
  const sym = String(symbol).trim().toUpperCase();
  const row = (results.symbol_results || []).find(
    (r) => String(r.symbol || '').toUpperCase() === sym,
  );
  if (!row) return { scoped: results, symbolError: null, missingSymbol: sym };
  if (row.error) return { scoped: null, symbolError: row.error };
  return {
    scoped: {
      ...results,
      total_pnl: row.total_pnl,
      trade_count: row.trade_count,
      summary: row.summary || {},
      walk_forward: row.walk_forward,
      _portfolioSymbol: sym,
    },
    symbolError: null,
  };
}

function validateWalkForwardOos(results, { minPnl, minTrades, minStabilityScore }) {
  const wf = results?.walk_forward || {};
  const oos = wf.out_of_sample || {};
  const summary = results?.summary || {};
  const aggregate = wf.aggregate || {};

  let oosPnl = oos.total_pnl;
  if (oosPnl == null) oosPnl = results?.total_pnl;
  oosPnl = Number(oosPnl || 0);

  let oosTrades = oos.trade_count ?? oos.summary?.total_trades;
  if (oosTrades == null) oosTrades = summary.total_trades ?? results?.trade_count;
  oosTrades = Number(oosTrades || 0);

  const foldCount = Number(aggregate.fold_count || 1);
  const stability = aggregate.stability_score;

  const metrics = {
    oos_pnl: oosPnl,
    oos_trades: oosTrades,
    stability_score: stability,
    fold_count: foldCount,
  };

  if (oosTrades < minTrades) {
    return {
      ok: false,
      reason: `OOS trades ${oosTrades} below minimum ${minTrades}`,
      metrics,
    };
  }
  if (oosPnl < minPnl) {
    return {
      ok: false,
      reason: `OOS PnL ${oosPnl.toFixed(2)} below minimum ${minPnl.toFixed(2)}`,
      metrics,
    };
  }
  if (
    minStabilityScore > 0
    && foldCount >= 3
    && stability != null
    && Number(stability) < minStabilityScore
  ) {
    return {
      ok: false,
      reason: `OOS stability ${(Number(stability) * 100).toFixed(0)}% below ${(minStabilityScore * 100).toFixed(0)}%`,
      metrics,
    };
  }
  return { ok: true, reason: 'OK', metrics };
}

function finalize(checks, workflowStage, metrics = {}) {
  const blocking = checks.some((c) => c.level === 'block' && !c.ok);
  const failed = checks.filter((c) => c.level === 'block' && !c.ok);
  return {
    passed: !blocking,
    blocking,
    checks,
    workflow_stage: workflowStage,
    block_reason: blocking ? (failed[0]?.message || 'Deploy gate blocked') : null,
    metrics,
  };
}

/**
 * Evaluate deploy prerequisites for the current backtest results + config.
 */
export function evaluateDeployGate({
  results,
  symbol,
  config,
  backtestConfig,
  snapshot,
  days,
  timeframe,
  strategy,
  thresholds = DEPLOY_GATE_DEFAULTS,
} = {}) {
  const {
    minTrades = DEPLOY_GATE_DEFAULTS.minTrades,
    minPnl = DEPLOY_GATE_DEFAULTS.minPnl,
    minStabilityScore = DEPLOY_GATE_DEFAULTS.minStabilityScore,
    maxDrawdownWarnPct = DEPLOY_GATE_DEFAULTS.maxDrawdownWarnPct,
  } = thresholds;

  const checks = [];

  if (!results) {
    checks.push(check({
      id: 'backtest_linked',
      level: 'warn',
      ok: false,
      message: 'No backtest loaded',
      detail: 'Run a backtest before deploying capital, or confirm deploy anyway.',
    }));
    return finalize(checks, 'backtest');
  }

  const { scoped, symbolError, missingSymbol } = symbolSlice(results, symbol);
  if (symbolError) {
    checks.push(check({
      id: 'symbol_backtest',
      level: 'block',
      ok: false,
      message: `Portfolio backtest failed for ${symbol}`,
      detail: symbolError,
    }));
    return finalize(checks, 'blocked');
  }

  let metrics = {};

  if (scoped?.walk_forward) {
    const { ok, reason, metrics: wfMetrics } = validateWalkForwardOos(scoped, {
      minPnl,
      minTrades,
      minStabilityScore,
    });
    metrics = wfMetrics;
    checks.push(check({
      id: 'wf_oos',
      level: ok ? 'pass' : 'block',
      ok,
      message: ok ? 'Walk-forward OOS validation passed' : reason,
      detail: ok ? null : reason,
    }));
    if (wfMetrics.stability_score != null && wfMetrics.fold_count >= 3) {
      const stab = Number(wfMetrics.stability_score);
      const stabOk = stab >= minStabilityScore;
      checks.push(check({
        id: 'wf_stability',
        level: stabOk ? 'pass' : 'block',
        ok: stabOk,
        message: stabOk
          ? `OOS stability ${(stab * 100).toFixed(0)}% across ${wfMetrics.fold_count} folds`
          : `OOS stability ${(stab * 100).toFixed(0)}% below ${(minStabilityScore * 100).toFixed(0)}%`,
      }));
    }
  } else {
    const summary = scoped?.summary || {};
    const pnl = Number(scoped?.total_pnl ?? summary.total_pnl ?? 0);
    const trades = Number(scoped?.trade_count ?? summary.total_trades ?? 0);
    const oosNote = scoped?.meta?.oos_pct
      ? `Results use ${scoped.meta.oos_pct}% OOS holdout window`
      : null;
    metrics = { pnl, trades };

    const tradesOk = trades >= minTrades;
    checks.push(check({
      id: 'trade_count',
      level: tradesOk ? 'pass' : 'block',
      ok: tradesOk,
      message: tradesOk
        ? `${trades} trades (minimum ${minTrades})`
        : `Only ${trades} trades — need at least ${minTrades}`,
      detail: oosNote,
    }));

    const pnlOk = pnl >= minPnl;
    checks.push(check({
      id: 'pnl',
      level: pnlOk ? 'pass' : 'block',
      ok: pnlOk,
      message: pnlOk
        ? `PnL $${pnl.toFixed(2)} meets minimum $${minPnl.toFixed(2)}`
        : `PnL $${pnl.toFixed(2)} below minimum $${minPnl.toFixed(2)}`,
      detail: oosNote,
    }));

    const maxDd = summary.max_drawdown_pct;
    if (maxDd != null && Number(maxDd) > maxDrawdownWarnPct) {
      checks.push(check({
        id: 'max_drawdown',
        level: 'warn',
        ok: false,
        message: `Max drawdown ${Number(maxDd).toFixed(1)}% exceeds ${maxDrawdownWarnPct}% guideline`,
      }));
    }
  }

  if (missingSymbol) {
    checks.push(check({
      id: 'portfolio_symbol',
      level: 'warn',
      ok: false,
      message: `No per-symbol result for ${missingSymbol}`,
      detail: 'Gate used aggregate portfolio metrics.',
    }));
  }

  const corr = results.correlation_summary;
  if (corr?.warning) {
    checks.push(check({
      id: 'correlation',
      level: 'warn',
      ok: false,
      message: 'High portfolio correlation',
      detail: corr.message,
    }));
  }

  if (snapshot && config) {
    const current = backtestFingerprint({
      symbol,
      strategy,
      days: String(days),
      timeframe,
      config,
      simMode: backtestConfig?.sim_mode ?? extractBacktestSimMode(results),
    });
    if (snapshot !== current) {
      checks.push(check({
        id: 'config_fingerprint',
        level: 'warn',
        ok: false,
        message: 'Config changed since last backtest',
        detail: 'Re-run backtest or deploy anyway to accept drift risk.',
      }));
    }
  }

  if (results && config) {
    const simMode = extractBacktestSimMode(results);
    if (simMode === 'research') {
      checks.push(check({
        id: 'research_sim_mode',
        level: 'warn',
        ok: false,
        message: 'Backtest used research mode',
        detail: 'Research allows shorts without live risk gates — re-run live-aligned before deploy, or confirm bypass.',
      }));
    }

    const deployDir = normalizeDirectionMode(config.direction_mode);
    const btDir = extractBacktestDirectionMode(results, backtestConfig);
    if (deployDir !== btDir) {
      checks.push(check({
        id: 'direction_mode_mismatch',
        level: 'warn',
        ok: false,
        message: `Trade direction mismatch: backtest ${formatDirectionModeLabel(btDir)}, deploy ${formatDirectionModeLabel(deployDir)}`,
        detail: 'Live bot risk gate uses deploy direction — shorts may be blocked or allowed differently than the backtest.',
      }));
    } else if (deployDir === 'BOTH' && simMode === 'live_aligned') {
      checks.push(check({
        id: 'direction_mode_both',
        level: 'pass',
        ok: true,
        message: 'Trade direction: both long & short (live-aligned)',
      }));
    }
  }

  let stage = 'ready';
  if (checks.some((c) => c.level === 'block' && !c.ok)) stage = 'blocked';
  else if (scoped?.walk_forward) stage = 'oos_validated';

  return finalize(checks, stage, metrics);
}

export function buildDeployPayload({
  strategy,
  symbol,
  timeframe,
  allocation,
  executionMode,
  config,
  results,
  snapshot,
  days,
  forceDeploy = false,
}) {
  const fingerprint = backtestFingerprint({
    symbol,
    strategy,
    days: String(days),
    timeframe,
    config,
    simMode: config?.sim_mode,
  });
  return {
    strategy,
    symbol,
    timeframe,
    allocation,
    execution_mode: executionMode,
    force_deploy: forceDeploy,
    backtest_fingerprint: snapshot || fingerprint,
    config: {
      ...config,
      trailing_stop_percent: config?.trailing_stop_percent ?? 2,
      backtest_run_id: results?.run_id ?? config?.backtest_run_id,
      backtest_fingerprint: snapshot || fingerprint,
    },
  };
}

File: backend/app/services/bots/deploy_gate.py
"""Deploy gate — forward-test before capital (backtest → OOS/WF → deploy)."""

from __future__ import annotations

import json
from datetime import datetime, timezone
from typing import Any

from app.services.agent.pipeline import validate_walk_forward_oos


def config_fingerprint(
    *,
    symbol: str | None = None,
    strategy: str | None = None,
    days: str | int | None = None,
    timeframe: str | None = None,
    config: dict | None = None,
) -> str:
    """Stable fingerprint for deploy parity (mirrors frontend backtestFingerprint)."""
    cfg = config or {}
    payload = {
        "symbol": symbol,
        "strategy": strategy,
        "days": str(days) if days is not None else None,
        "timeframe": timeframe,
        "allocation": cfg.get("allocation"),
        "trailing_stop_percent": cfg.get("trailing_stop_percent"),
        "take_profit_percent": cfg.get("take_profit_percent"),
        "tp_mode": cfg.get("tp_mode"),
        "min_confidence": cfg.get("min_confidence"),
    }
    return json.dumps(payload, sort_keys=True, separators=(",", ":"))


def _check(
    *,
    check_id: str,
    level: str,
    ok: bool,
    message: str,
    detail: str | None = None,
) -> dict[str, Any]:
    return {
        "id": check_id,
        "level": level,
        "ok": ok,
        "message": message,
        "detail": detail,
    }


def _symbol_slice(results: dict, symbol: str | None) -> dict:
    """Portfolio runs: evaluate the row for the deploy symbol when possible."""
    if not results.get("portfolio") or not symbol:
        return results
    sym = str(symbol).upper()
    for row in results.get("symbol_results") or []:
        if str(row.get("symbol") or "").upper() == sym:
            if row.get("error"):
                return {**results, "_symbol_error": row.get("error")}
            return {
                **results,
                "total_pnl": row.get("total_pnl"),
                "trade_count": row.get("trade_count"),
                "summary": row.get("summary") or {},
                "walk_forward": row.get("walk_forward"),
                "_portfolio_symbol": sym,
            }
    return results


def evaluate_deploy_gate(
    results: dict | None,
    *,
    symbol: str | None = None,
    deploy_fingerprint: str | None = None,
    run_config: dict | None = None,
    run_days: str | int | None = None,
    run_timeframe: str | None = None,
    min_trades: int = 1,
    min_pnl: float = 0.0,
    min_stability_score: float = 0.5,
    max_drawdown_warn_pct: float = 25.0,
) -> dict[str, Any]:
    """Evaluate whether a backtest run satisfies deploy prerequisites."""
    checks: list[dict[str, Any]] = []

    if not results:
        checks.append(_check(
            check_id="backtest_linked",
            level="warn",
            ok=False,
            message="No backtest run linked",
            detail="Run a backtest and deploy from results, or use force deploy.",
        ))
        return {
            "passed": True,
            "blocking": False,
            "checks": checks,
            "workflow_stage": "backtest",
            "block_reason": None,
            "metrics": {},
        }

    scoped = _symbol_slice(results, symbol)
    sym_err = scoped.pop("_symbol_error", None)
    if sym_err:
        checks.append(_check(
            check_id="symbol_backtest",
            level="block",
            ok=False,
            message=f"Portfolio backtest failed for {symbol}",
            detail=str(sym_err),
        ))
        return _finalize(checks, workflow_stage="blocked")

    metrics: dict[str, Any] = {}

    if scoped.get("walk_forward"):
        ok, reason, metrics = validate_walk_forward_oos(
            scoped,
            min_oos_pnl=min_pnl,
            min_oos_trades=min_trades,
            min_stability_score=min_stability_score,
        )
        checks.append(_check(
            check_id="wf_oos",
            level="block" if not ok else "pass",
            ok=ok,
            message="Walk-forward OOS validation passed" if ok else reason,
            detail=None if ok else reason,
        ))
        if metrics.get("stability_score") is not None and int(metrics.get("fold_count") or 0) >= 3:
            stab = float(metrics["stability_score"])
            stab_ok = stab >= min_stability_score
            checks.append(_check(
                check_id="wf_stability",
                level="block" if not stab_ok else "pass",
                ok=stab_ok,
                message=(
                    f"OOS stability {stab:.0%} across {metrics['fold_count']} folds"
                    if stab_ok
                    else f"OOS stability {stab:.0%} below {min_stability_score:.0%}"
                ),
            ))
    else:
        summary = scoped.get("summary") or {}
        pnl = scoped.get("total_pnl")
        if pnl is None:
            pnl = summary.get("total_pnl")
        try:
            pnl_f = float(pnl or 0)
        except (TypeError, ValueError):
            pnl_f = 0.0

        trades = scoped.get("trade_count")
        if trades is None:
            trades = summary.get("total_trades")
        try:
            trades_i = int(trades or 0)
        except (TypeError, ValueError):
            trades_i = 0

        meta = scoped.get("meta") or {}
        oos_note = None
        if meta.get("oos_pct"):
            oos_note = f"Results use {meta['oos_pct']}% OOS holdout window"

        trades_ok = trades_i >= max(0, int(min_trades))
        checks.append(_check(
            check_id="trade_count",
            level="block" if not trades_ok else "pass",
            ok=trades_ok,
            message=(
                f"{trades_i} trades (minimum {min_trades})"
                if trades_ok
                else f"Only {trades_i} trades — need at least {min_trades}"
            ),
            detail=oos_note,
        ))

        pnl_ok = pnl_f >= float(min_pnl)
        checks.append(_check(
            check_id="pnl",
            level="block" if not pnl_ok else "pass",
            ok=pnl_ok,
            message=(
                f"PnL ${pnl_f:.2f} meets minimum ${min_pnl:.2f}"
                if pnl_ok
                else f"PnL ${pnl_f:.2f} below minimum ${min_pnl:.2f}"
            ),
            detail=oos_note,
        ))
        metrics = {"pnl": round(pnl_f, 4), "trades": trades_i}

        max_dd = summary.get("max_drawdown_pct")
        if max_dd is not None:
            try:
                dd_f = float(max_dd)
                if dd_f > max_drawdown_warn_pct:
                    checks.append(_check(
                        check_id="max_drawdown",
                        level="warn",
                        ok=False,
                        message=f"Max drawdown {dd_f:.1f}% exceeds {max_drawdown_warn_pct:.0f}% guideline",
                    ))
            except (TypeError, ValueError):
                pass

    if scoped.get("portfolio") and not scoped.get("_portfolio_symbol") and symbol:
        checks.append(_check(
            check_id="portfolio_symbol",
            level="warn",
            ok=False,
            message=f"No per-symbol result for {symbol} in portfolio run",
            detail="Gate used aggregate portfolio metrics.",
        ))

    corr = scoped.get("correlation_summary") or results.get("correlation_summary")
    if corr and corr.get("warning"):
        checks.append(_check(
            check_id="correlation",
            level="warn",
            ok=False,
            message="High portfolio correlation",
            detail=corr.get("message"),
        ))

    if deploy_fingerprint and run_config is not None:
        run_fp = config_fingerprint(
            symbol=symbol or results.get("meta", {}).get("symbol"),
            strategy=results.get("meta", {}).get("strategy"),
            days=run_days,
            timeframe=run_timeframe,
            config=run_config,
        )
        if run_fp != deploy_fingerprint:
            checks.append(_check(
                check_id="config_fingerprint",
                level="warn",
                ok=False,
                message="Deploy config differs from backtest snapshot",
                detail="Strategy parameters changed since the linked backtest run.",
            ))

    event_manifest = (scoped.get("meta") or results.get("meta") or {}).get("event_manifest") or {}
    splits = int(event_manifest.get("splits_in_range") or 0)
    price_adjust = str(event_manifest.get("price_adjust") or "raw")
    if splits > 0 and price_adjust == "raw":
        checks.append(_check(
            check_id="corp_split_adjust",
            level="warn",
            ok=False,
            message=f"Backtest spans {splits} split(s) with raw (unadjusted) prices",
            detail="Set BACKTEST_PRICE_ADJUST=split_only or total_return for accurate PnL.",
        ))

    stage = "ready"
    if any(c["level"] == "block" and not c["ok"] for c in checks):
        stage = "blocked"
    elif any(c["id"] == "backtest_linked" for c in checks):
        stage = "backtest"
    elif scoped.get("walk_forward"):
        stage = "oos_validated" if stage == "ready" else stage

    return _finalize(checks, workflow_stage=stage, metrics=metrics)


def _finalize(
    checks: list[dict[str, Any]],
    *,
    workflow_stage: str,
    metrics: dict | None = None,
) -> dict[str, Any]:
    blocking = any(c["level"] == "block" and not c["ok"] for c in checks)
    passed = not blocking
    block_reason = None
    if blocking:
        failed = [c for c in checks if c["level"] == "block" and not c["ok"]]
        block_reason = failed[0]["message"] if failed else "Deploy gate blocked"
    return {
        "passed": passed,
        "blocking": blocking,
        "checks": checks,
        "workflow_stage": workflow_stage,
        "block_reason": block_reason,
        "metrics": metrics or {},
    }


def enrich_deploy_config(
    config: dict | None,
    *,
    run_id: str | None = None,
    fingerprint: str | None = None,
    gate: dict | None = None,
) -> dict:
    """Persist deploy audit fields on the bot config."""
    out = dict(config or {})
    if run_id:
        out["backtest_run_id"] = run_id
    if fingerprint:
        out["backtest_fingerprint"] = fingerprint
    if gate and gate.get("passed"):
        out["deploy_gate_passed_at"] = datetime.now(timezone.utc).isoformat()
    out.setdefault("deploy_workflow", "backtest→oos→deploy")
    return out
